gaf/docs/gaf.1.in
.TH gaf 1 "@DATE@" "gEDA Project" @VERSION@
.SH NAME
gaf - gEDA/gaf Command-Line Utility
.SH "SYNOPSIS"
.B gaf
[\fIOPTION\fR ...] \fICOMMAND\fR [\fIARGS\fR ...]
.SH "DESCRIPTION"
.PP
.B gaf
is part of the gEDA (GPL Electronic Design Automation) toolset.  It
provides a number of small command-line utilities for working with
schematic and symbol files, and is designed to be used for batch
processing of designs created using the schematic editor
\fBgschem\fR(1).  It currently has four built-in \fICOMMAND\fRs:

.B gaf export
is used to create SVG, PDF, PNG, PS and EPS files from schematic and
symbol files, for printing or embedding in other documents.

.B gaf config
allows reading and writing settings in gEDA project, user and system
configuration stores.

.B gaf shell
provides a Scheme REPL for command-line batch processing of schematic
data.

.B gaf path
Writes path strings. This can be useful in scripts requiring this information.

.SH "GENERAL OPTIONS"
.TP 8
\fB-n\fR, \fB--no-rcfiles\fR
Prevent `gafrc' Scheme initialization files from being loaded.
.TP 8
\fB-h\fR, \fB--help\fR
Print a help message.
.TP 8
\fB-V\fR, \fB--version\fR
Print \fBgaf\fR version information.

.SH "EXPORTING IMAGE FILES"
.B gaf export
[\fIOPTION\fR ...] \fB-o\fR \fIOUTPUT\fR [\fI--\fR] \fIFILE\fR ...

.B gaf export
can export schematic and symbol files in a variety of image formats
for printing or further processing. Export currently supports single-page
PNG, SVG and EPS output, and multi-page PS and PDF output. The following
options can be used to control how the output is formatted.

.TP 8
\fB-o\fR, \fB--output\fR=\fIFILE\fR
Output generated image data to \fIFILE\fR.
.TP 8
\fB-f\fR, \fB--format\fR=(\fBpng\fR | \fBpdf\fR | \fBsvg\fR | \fBps\fR | \fBeps\fR)
Specify an output format.  Usually, this option is not required,
because \fBgaf export\fR will infer the correct format from the file
extension of the output file.
.TP 8
\fB-p\fR, \fB--paper\fR=\fINAME\fR
Size the output for a particular paper size.  The \fINAME\fR should be
a PWG 5101.1-2002 paper name.  For example, valid values include
`iso_a4', `na_letter', or `na_d'.
.TP 8
\fB-l\fR, \fB--layout\fR=(\fBauto\fR | \fBlandscape\fR | \fBportrait\fR)
When using a paper size, set the orientation of the output.  If `auto'
layout is used, the orientation that best fits the drawing will be
used.
.TP 8
\fB-s\fR, \fB--size\fR=(\fBauto\fR | \fIWIDTH\fR:\fIHEIGHT\fR)
Size the output with specific dimensions. If the size is `auto', select the size
that best fits the drawing. A comma, semi-colon, or space can also be used to
separate the \fIWIDTH\fR and \fIHEIGHT\fR but parenthesis will likely be required
if a semi colon or a space is used as the separator.
.TP 8
\fB-k\fR, \fB--scale\fR=\fIFACTOR\fR
Set the output scale \fIFACTOR\fR. This is a distance identical with 100
points (1 default grid spacing) in \fBgschem\fR(1) coordinate space. It is used
to size the output when neither \fB--paper\fR nor \fB--size\fR are
given, and defaults to 100 mil.
.TP 8
\fB-m\fR, \fB--margins\fR=(\fBauto\fR | \fITOP\fR[:\fILEFT\fR[:\fIBOTTOM\fR[:\fIRIGHT\fR]]])
Set the widths of the margins to be used.  If `auto' margins are
specified, a sensible default value will be chosen.  Up to four margin
widths can be provided.  If one is provided, it will be used on all
four sides.  If two are provided, the first will be used for the
top/bottom and the second for the left/right.  If three are provided,
the first will be used for the top, the second for left/right, and the
third for the bottom.
.TP 8
\fB-a\fR, \fB--align\fR=(\fBauto\fR | \fIHALIGN\fR:\fIVALIGN\fR)
Set how the drawing is aligned within the page.  \fIHALIGN\fR controls
the horizontal alignment, and \fIVALIGN\fR the vertical.  Each
alignment value should be in the range 0.0 to 1.0.  The `auto'
alignment is equivalent to a value of `0.5:0.5', i.e. centered.
.TP 8
\fB-d\fR, \fB--dpi\fR=\fIDPI\fR
Set the number of pixels per inch used when generating PNG output.
.TP 8
\fB-c\fR, \fB--color\fR
Enable colour output.
.TP 8
\fB--no-color\fR
Disable color output.
.TP 8
\fB-F\fR, \fB--font\fR=\fIFONT-FAMILY\fR
Set the font to be used for drawing text.
.TP 8
\fB--\fR
Treat all remaining arguments as schematic or symbol filenames.  Use
this if you have a schematic or symbol filename which begins with `-'.

.PP
The \fB--size\fR, \fB--margins\fR, or \fB--scale\fR options described above accept
values using units of `mm', `cm', `in', `pc', `px', or `pt'.  If you
do not provide a unit, points are assumed.  N.b. that `px' are
evaluated relative to the current \fB--dpi\fR setting.

.PP
When using the \fB--size\fR, \fB--margins\fR, or \fB--align\fR options
with multiple values, you may use `;', or ` ' as a separator between
them instead of `:'. In such a case, remember to properly quote your
arguments to avoid them to be interpreted by your shell.


.SH "ACCESSING CONFIGURATION"
.B gaf config
[\fIOPTION\fR] [\fIGROUP\fR \fIKEY\fR [\fIVALUE\fR]]

.B gaf config
is a basic utility for inspecting and modifying gEDA/gaf configuration
stores.

.PP
If a \fIGROUP\fR and \fRKEY\fR are specified, retrieves the value of
that configuration parameter.  If a \fIVALUE\fR is specified, sets
the value of that parameter.  The \fB-p\fR, \fB-u\fR and \fB-s\fR
options can be used to select the configuration store affected (by
default, the project configuration store for the current working
directory).  If no \fIGROUP\fR and \fIKEY\fR are provided, outputs
the filename of the selected configuration store.

.PP
.TP 8
\fB-p\fR, \fB--project\fR[=\fIPATH\fR]
Select the project configuration store of the project located in
\fIPATH\fR.  If no \fIPATH\fR is provided, selects the project in the
current working directory.
.TP 8
\fB-u\fR, \fB--user\fR
Select the user configuration store.
.TP 8
\fB-s\fR, \fB--system\fR
Select the system configuration store.  Depending on user permissions,
the system configuration store may be read-only.

.SH "SCHEME PROCESSING"
.B gaf shell
[\fIOPTION ...\fR]

.B gaf shell
provides a Scheme Read-Eval-Print Loop (REPL) for automating
processing of schematic and symbol files.  It is designed to be used
with the gEDA Scheme API.

.TP 8
\fB-e\fR , \fB--eval \fIEXPR\fR
Evaluate the Scheme expression \fIEXPR\fR, and exit.
.TP 8
\fB-h\fR, \fB--help\fR
Displays help information on options specific to the shell back-end.
.TP 8
\fB\-l\fR, \fB\--load \fIFILE\fR
Evaluate Scheme source code from \fIFILE\fR.
.TP 8
\fB\-p\fR, \fB\--path DIRECTORY\fR
Prepend \fIDIRECTORY\fR to the list of directories to be searched for Scheme files.
\fB-s\fR \fB\--scheme \fIFILE\fR
Evaluate Scheme source code from \fIFILE\fR, and exit.
.TP 8
\fB--\fR
Stop scanning arguments; run interactively.

.PP
The \fB-e\fR, \fB-s\fR and \fB--\fR switches stop argument processing
and pass all the remaining arguments as the value of `(command-line)'.

.SH "OBTAINING PATH INFORMATION"
.B gaf path
[\fIOPTION\fR] [\fIWHICH ONE\fR]]

.B gaf path
Allows scripts other than scheme to obtain the location of gEDA directories from libgeda.
Scheme scripts have access to this information but it can be inconvenient for other scripting
languages to access this information through scheme.
.TP 8
example:
datapath=$(gaf path -d)

.SH ENVIRONMENT
.TP 8
.B GEDADATA
specifies the search directory for Scheme and rc files.  The default
is `${prefix}/share/gEDA'.
.TP 8
.B GEDADATARC
specifies the search directory for rc files.  The default is `$GEDADATA'.

.SH AUTHORS
See the `AUTHORS' file included with this program.

.SH COPYRIGHT
.nf
Copyright \(co 2012-@YEAR@ gEDA Contributors.  License GPLv2+: GNU GPL
version 2 or later.  Please see the "COPYING" file included with this
program for full details.
.PP
This is free software: you are free to change and redistribute it. There is NO
WARRANTY; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE, to
the extent permitted by law.

.SH SEE ALSO
\fBgschem\fR(1).

For more information on the Scheme API, see the \fBgeda-scheme\fR
Texinfo manual.  If the \fBinfo\fR program is properly installed at
your site, the command

.IP
.B
info geda-scheme

.PP
should give you access to the complete manual.


NEWS
==============================================
 'gschem and Friends' Electronic Design Suite
==============================================

Copyright (C) 1998-2017 gEDA Developers

This file documents important user-visible changes in gEDA/gaf.  For
more information, please consult the `ChangeLog' file.

Notable changes in gEDA/gaf gEDA 2.1.1.1 - 11/12/17 Rampage Edition

* Expanded support for internationalization (i18n).

* The build system now support using the --with-included-gettext option to
  configure the suite to build and use an in-source international library.

* Font-sizes for widgets on GedaToolbars can now be specified in RC files.

* The toolbar layout in gschem has been improved to better support labels.

* The visibility of toolbar tooltips can now be control using RC keywords.

* The Slot Editor Dialog now incorporates a spinner control to adjust the
  slot number of slotted components. The range is restricted to valid slots
  for the selected component.

* New bitmaps have been added;

      * geda-deslect.png
      * geda-find-attribute
      * geda-properties.png
      * geda-trash-can.png

* libgedathon no longer over-writes existing gafrc when the declare_local_sym
  feature is used.

* libgedathon now supports internationalization (i18n).

Notable changes in gEDA/gaf gEDA 2.1.1.0 - 01/01/17 Rampage Edition

* Improvements were made to both the build system and source code to ensure
  gEDA-GAF builds more reliably on 64 bit platforms. Testing was performed
  using Ubuntu 12.04 and Ubuntu 14.04 64 bit host systems.

* The libgeda API has been completely redone in order to present a library API
  that is not only for in-suite applications.

* Support for internationalization (i18n) has been significantly expanded and
  now includes command-line utilities.

* Extensive revisions have been made to improve performance of gEDA application.
  This is particularly noticeable in gschem.

* A new library, libgedacolor, now unifies color management for the gEDA suite.
  Color-maps are still configured using scheme files, but the use of "friendly-
  color-maps"  is no longer applicable; the new color maps for libgedacolor are
  already "friendly". The library was added to provide a unified API for gEDA
  applications and should be transparent to most users. User who are using
  custom color-maps will need to modify their custom map files.

* A new auto-panning feature has been added to gschem. The auto-pan is only in
  effect while drawing nets and buses and can enabled or disabled through menu
  options.

* The Auto-Number dialog in gschem no longer requires the use of wild-card
  characters in the scope text. This is a new advanced feature that allows
  automatic re-numbering of duplicated reference designators.

* The path for logs file can now be set in initialization files, including
  local gafrc files. This allows logs to be organized on a per project basis.
  The location for logs can also be set using a new GEDALOGS environment variable.

* A new libgeda keyword "log-time" can be use to enable or disable prefixing
  log entries with the time of day.

* The Multi-Attribute Editor Dialog entry now has attribute "word" completion
  identical to the Single-Attribute Editor Dialog.

* The list of valid attribute used by gsymcheck when checking symbol files is
  now users configurable. The list can be set in system-gsymcheckrc.

* The list of known devices recognized by gsymcheck when checking the device
  entries in symbol files is now users configurable. The list can be set in
  system-gsymcheckrc.

* Symbol files can now be renamed using the Component Dialog.

* The number of user defined colors was increased by 6 to a total of
  10 user definable colors in each color map.

* The main menu system has been modified to automatically build a basic
  fall-back menu when errors are discovered in the menu RC file.

* The height of the status bar in Gschem can now be set in the gschem-gtkrc

* HotKey dialog now includes actions icon and a column to display the actual
  binding string. User can enable or disable viewing the action column with
  the setting being retained between session.

* The Slot Editor Dialog now displays the total number of slots per component.

* New features have been added to the Page Select Dialog:

      Accept drag-and-drop of schematic or symbol files.
      Opened pages can be re-ordered either up or down.
      Auto scrolls to selection.
      Shows long filename as tooltip when showing short names.
      Optionally adjust height of dialog automatically.

* Exchange image file now retains the user preference to update all selected
  images between session. Previously the user preference was only retained for
  current session.

* Additional preference variables have been incorporated in new the configuration
  system allow retention of changes using the Preferences dialog:

      continue-component-place
      embed-components
      enforce-hierarchy
      force-boundingbox
      keyboardpan-gain
      netconn-rubberband
      select-slack-pixels
      snap-size
      sort-component-library
      object-clipping

  Uses of these keywords in rc files is still permissible and will over-ride the
  preference setting.

* A new gschem command line options has been provided to specify a title-block at
  startup.

* A new gschem command line options has been provided to startup minimized.

* The command line option to add a directory to the Guile path now has a long option.
  The "-L" option has been replaced with "-g" or "--guile".

* gsymcheck now supports the standard command-line version options.

* Programs in the geda-gaf suite now support quite mode in conjunction with the version
  option to assist with extraction of dotted version information.

* New Scheme API have been added:

      gschem-bmp-image
      gschem-jpeg-image
      gschem-tiff-image

* The maximum zoom magnification in gschem has been increased slightly.


gaf/src/gaf.c
/* -*- Mode: C; indent-tabs-mode: nil; c-basic-offset: 2 tab-width: 4 -*- */
/* vi: set et ts=4 sw=2 sts=2: */
/*
 * gEDA/gaf command-line utility
 * Copyright (C) 2012-2016 Peter Brett <peter@peter-b.co.uk>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301 USA <http://www.gnu.org/licenses/>.
 */

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include <version.h>

#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <getopt.h>

#include <glib.h>

/* Gettext translation */
#include <locale.h>
#include "../include/gettext.h"
#include "../include/builtins.h"

#include <libgeda/libgeda.h>         /* for geda_sprintf */

#define short_options "+hnV"

static struct option long_options[] =
  {
    {"help", 0, NULL, 'h'},
    {"no-rcfiles", 0, NULL, 'n'},
    {"version", 0, NULL, 'V'},
    {NULL, 0, NULL, 0},
  };

struct internal_command {
  char *name;
  int (*func)(int, char **);
};

static struct internal_command commands[] =
  {
    {"shell", cmd_shell},
    {"config", cmd_config},
    {"export", cmd_export},
    {"path", cmd_path},
    {NULL, NULL},
  };

/* Print help info and exit */
static void usage (void)
{
  printf (_("Usage: gaf [OPTION...] COMMAND [ARGS ...]\n\n"
            "gEDA/gaf command-line utility.\n\n"
            "General options:\n"
            "  -n, --no-rcfiles  inhibit loading of 'gafrc' files\n"
            "  -h, --help        display usage information and exit\n"
            "  -V, --version     display version information and exit\n\n"
            "Commonly-used commands (type `gaf <cmd> --help' for usage):\n"
            "  shell             Scheme REPL for interactive gEDA data processing\n"
            "  config            Edit gEDA configuration\n"
            "  export            Export gEDA files in various image formats.\n"
            "  path              Writes path strings to standard output port.\n\n"
            "Please report bugs to %s.\n"),
             PACKAGE_BUGREPORT);
  exit (0);
}

/* Print version info and exit */
static void version (void)
{

  printf(_("gEDA/gaf %s (%s) (g%.7s)\n"
           "Copyright (C) 1998-2017 gEDA developers\n"
           "This is free software, and you are welcome to redistribute it under\n"
           "certain conditions. For details, see the file `COPYING', which is\n"
           "included in the gEDA distribution.\n"
           "There is NO WARRANTY, to the extent permitted by law.\n"),
           PACKAGE_DOTTED_VERSION, PACKAGE_DATE_VERSION, PACKAGE_GIT_COMMIT);
  exit (0);
}

static void gaf_show_run_help (const char *msg)
{
  const char *help_msg = _("Run `gaf --help' for more information");

  if (msg) {
    fprintf (stderr, "\n%s.\n\n%s.\n\n", msg, help_msg);
  }
  else {
    fprintf (stderr, "\n%s.\n\n", help_msg);
  }
}

int main (int argc, char **argv)
{
  int    c;
  char  *cmd;
  int    cmd_argc;
  char **cmd_argv;

  int (*cmd_func)(int, char **) = NULL;

  /* Possibly set up gettext */
#if ENABLE_NLS

  setlocale (LC_ALL, "");
  setlocale (LC_NUMERIC, "C");
  bindtextdomain ("geda-gaf", LOCALEDIR);
  textdomain ("geda-gaf");
  bind_textdomain_codeset ("geda-gaf", "UTF-8");

#endif

  while (-1 != (c = getopt_long (argc, argv, short_options,
                                 long_options, NULL))) {
    switch (c) {

    case 0:
      /* This is a long-form-only flag option, and has already been
       * dealt with by getopt_long(). */
      break;

    case 'V':
      version ();

    case 'h':
      usage ();

    case 'n':
      g_setenv ("GAF_INHIBIT_RCFILES", "1", 1);
      break;

    case '?':
      /* getopt_long already printed an error message */
      gaf_show_run_help (NULL);
      exit (1);
      break;

    default:
      fprintf (stderr, "%s: unhandled case <%d>\n", __func__, c);
    }
  }

  /* The next argument should be a command */
  if (optind == argc) {
    gaf_show_run_help (_("No command specified, nothing to do"));
    exit (1);
  }

  cmd = argv[optind];

  /* Look up the command */
  int i;
  for (i = 0; commands[i].name != NULL; i++) {
    if (strcmp (cmd, commands[i].name) == 0) {
      cmd_func = commands[i].func;
      break;
    }
  }

  if (cmd_func == NULL) {

    const char *msg = _("ERROR: Unrecognised command");
          char *str;

    str = geda_sprintf ("%s <%s>", msg, cmd);

    gaf_show_run_help (str);

    free(str);
    exit (1);
  }

  cmd_argc = argc - optind;
  cmd_argv = argv + optind;
  optind = 1;

  return cmd_func (cmd_argc, cmd_argv);
}
